openprescribing/frontend/tests/price_per_unit/test_savings.py

```diff
@@ -0,0 +1,320 @@
+from collections import defaultdict
+import json
+import warnings
+
+import numpy
+
+from django.core.cache import CacheKeyWarning
+from django.test import TestCase, override_settings
+
+from matrixstore.tests.data_factory import DataFactory
+from matrixstore.tests.matrixstore_factory import (
+    patch_global_matrixstore,
+    matrixstore_from_data_factory,
+)
+from frontend.models import Practice, PCT, Presentation
+from frontend.price_per_unit.substitution_sets import get_substitution_sets
+from frontend.price_per_unit.savings import (
+    CONFIG_MIN_SAVINGS_FOR_ORG_TYPE,
+    CONFIG_TARGET_CENTILE,
+    get_total_savings_for_org,
+)
+
+
+# The dummy cache backend we use in testing warns that our binary cache keys
+# won't be compatible with memcached, but we really don't care
+warnings.simplefilter("ignore", CacheKeyWarning)
+DUMMY_CACHE_SETTING = {
+    "default": {"BACKEND": "django.core.cache.backends.dummy.DummyCache"}
+}
+
+
+@override_settings(CACHES=DUMMY_CACHE_SETTING)
+class PricePerUnitSavingsTest(TestCase):
+    maxDiff = None
+
+    @classmethod
+    def setUpTestData(cls):
+        # Create some sets of substitutable presentations, which each consist
+        # of a generic plus some branded equivalents
+        substitution_sets = []
+        for i in range(4):
+            generic_code = invent_generic_bnf_code(i)
+            brands = invent_brands_from_generic_bnf_code(generic_code)
+            substitution_sets.append([generic_code] + brands)
+
+        # Create some practices and some prescribing for these presentations
+        factory = DataFactory()
+        factory.create_months("2020-01-01", 2)
+        factory.create_practices(10)
+        for bnf_codes in substitution_sets:
+            for bnf_code in bnf_codes:
+                factory.create_presentation(bnf_code=bnf_code)
+        factory.create_prescribing(
+            factory.presentations, factory.practices, factory.months
+        )
+
+        # Copy these objects to the database
```

Comment:
I don't understand this comment.

Reply:
I've expanded this comment:
```
# The DataFactory creates data that can be written to the MatrixStore
# but doesn't automatically create anything in the database, so we do
# that manually here
```
But maybe the route of the confusion is that we've ended up with two DataFactory classes: one that lives in `matrixstore.tests` (which deliberately avoids writing to the database) and another that lives in `frontend.tests` which automatically populates the database.

It would be good to unify these, but it's a slightly tricky and unrewarding job I fear.